// @liveblocks/react/dist/_private.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }

























var _chunkUR7FENGRcjs = require('./chunk-UR7FENGR.cjs');

// src/lib/use-layout-effect.ts
var _react = require('react');
var useLayoutEffect = typeof window !== "undefined" ? _react.useLayoutEffect : _react.useEffect;

// src/use-mention-suggestions.ts
var _core = require('@liveblocks/core');

var MENTION_SUGGESTIONS_DEBOUNCE = 500;
function normalizeMentionSuggestions(suggestions) {
  return suggestions.map(
    (suggestion) => typeof suggestion === "string" ? { kind: "user", id: suggestion } : suggestion
  );
}
function useMentionSuggestions(roomId, search) {
  const [mentionSuggestions, setMentionSuggestions] = _react.useState.call(void 0, );
  const lastInvokedAt = _react.useRef.call(void 0, );
  const resolveMentionSuggestions = _chunkUR7FENGRcjs.useResolveMentionSuggestions.call(void 0, );
  const mentionSuggestionsCache = _chunkUR7FENGRcjs.useMentionSuggestionsCache.call(void 0, );
  _react.useEffect.call(void 0, () => {
    if (search === void 0 || !resolveMentionSuggestions) {
      return;
    }
    const resolveMentionSuggestionsArgs = { text: search, roomId };
    const mentionSuggestionsCacheKey = _core.stableStringify.call(void 0, 
      resolveMentionSuggestionsArgs
    );
    let debounceTimeout;
    let isCanceled = false;
    const getMentionSuggestions = async () => {
      try {
        lastInvokedAt.current = performance.now();
        const rawMentionSuggestions = await resolveMentionSuggestions(
          resolveMentionSuggestionsArgs
        );
        if (!isCanceled) {
          const normalizedSuggestions = normalizeMentionSuggestions(
            rawMentionSuggestions
          );
          setMentionSuggestions(normalizedSuggestions);
          mentionSuggestionsCache.set(
            mentionSuggestionsCacheKey,
            normalizedSuggestions
          );
        }
      } catch (error) {
        console.error(_optionalChain([error, 'optionalAccess', _ => _.message]));
      }
    };
    if (mentionSuggestionsCache.has(mentionSuggestionsCacheKey)) {
      const cachedSuggestions = mentionSuggestionsCache.get(
        mentionSuggestionsCacheKey
      );
      setMentionSuggestions(cachedSuggestions);
    } else if (!lastInvokedAt.current || Math.abs(performance.now() - lastInvokedAt.current) > MENTION_SUGGESTIONS_DEBOUNCE) {
      void getMentionSuggestions();
    } else {
      debounceTimeout = window.setTimeout(() => {
        void getMentionSuggestions();
      }, MENTION_SUGGESTIONS_DEBOUNCE);
    }
    return () => {
      isCanceled = true;
      window.clearTimeout(debounceTimeout);
    };
  }, [search, roomId, resolveMentionSuggestions, mentionSuggestionsCache]);
  return mentionSuggestions;
}

// src/use-sync-source.ts


function useSyncSource() {
  const client = _chunkUR7FENGRcjs.useClient.call(void 0, );
  const createSyncSource = client[_core.kInternal].createSyncSource;
  const [syncSource, setSyncSource] = _react.useState.call(void 0, );
  _react.useEffect.call(void 0, () => {
    const newSyncSource = createSyncSource();
    setSyncSource(newSyncSource);
    return () => newSyncSource.destroy();
  }, [createSyncSource]);
  return syncSource;
}




























exports.getUmbrellaStoreForClient = _chunkUR7FENGRcjs.getUmbrellaStoreForClient; exports.useAddRoomCommentReaction = _chunkUR7FENGRcjs.useAddRoomCommentReaction; exports.useClientOrNull = _chunkUR7FENGRcjs.useClientOrNull; exports.useCreateRoomComment = _chunkUR7FENGRcjs.useCreateRoomComment; exports.useCreateRoomThread = _chunkUR7FENGRcjs.useCreateRoomThread; exports.useCreateTextMention = _chunkUR7FENGRcjs.useCreateTextMention; exports.useDeleteRoomComment = _chunkUR7FENGRcjs.useDeleteRoomComment; exports.useDeleteRoomThread = _chunkUR7FENGRcjs.useDeleteRoomThread; exports.useDeleteTextMention = _chunkUR7FENGRcjs.useDeleteTextMention; exports.useEditRoomComment = _chunkUR7FENGRcjs.useEditRoomComment; exports.useEditRoomThreadMetadata = _chunkUR7FENGRcjs.useEditRoomThreadMetadata; exports.useLayoutEffect = useLayoutEffect; exports.useMarkRoomThreadAsRead = _chunkUR7FENGRcjs.useMarkRoomThreadAsRead; exports.useMarkRoomThreadAsResolved = _chunkUR7FENGRcjs.useMarkRoomThreadAsResolved; exports.useMarkRoomThreadAsUnresolved = _chunkUR7FENGRcjs.useMarkRoomThreadAsUnresolved; exports.useMentionSuggestions = useMentionSuggestions; exports.useMentionSuggestionsCache = _chunkUR7FENGRcjs.useMentionSuggestionsCache; exports.useRemoveRoomCommentReaction = _chunkUR7FENGRcjs.useRemoveRoomCommentReaction; exports.useReportTextEditor = _chunkUR7FENGRcjs.useReportTextEditor; exports.useResolveMentionSuggestions = _chunkUR7FENGRcjs.useResolveMentionSuggestions; exports.useRoomAttachmentUrl = _chunkUR7FENGRcjs.useRoomAttachmentUrl; exports.useRoomPermissions = _chunkUR7FENGRcjs.useRoomPermissions; exports.useRoomThreadSubscription = _chunkUR7FENGRcjs.useRoomThreadSubscription; exports.useSignal = _chunkUR7FENGRcjs.useSignal; exports.useSyncExternalStoreWithSelector = _chunkUR7FENGRcjs.useSyncExternalStoreWithSelector; exports.useSyncSource = useSyncSource; exports.useYjsProvider = _chunkUR7FENGRcjs.useYjsProvider;
